cosmo/log.py

```diff
@@ -1,2 +1,166 @@
+from abc import abstractmethod, ABCMeta
+from typing import Self
+
+from cosmo.common import AbstractRecoverableError, JsonOutputType
+from cosmo.netbox_types import AbstractNetboxType
+
+
+class AbstractLogLevel(metaclass=ABCMeta):
+    name: str = "abstract_log_level"
+    def __str__(self):
+        return f"[{self.name.upper()}]"
+
+class InfoLogLevel(AbstractLogLevel):
+    name = "info"
+class WarningLogLevel(AbstractLogLevel):
+    name = "warning"
+class ErrorLogLevel(AbstractLogLevel):
+    name = "error"
+
+
+O = object | AbstractNetboxType | None # object-being-logged-on type
+M = tuple[AbstractLogLevel, str, O] # message type
+
+class AbstractLoggingStrategy(metaclass=ABCMeta):
+    @abstractmethod
+    def flush(self): # this is for async and sync logging
+        pass
+    @abstractmethod
+    def info(self, message: str, on: O):
+        pass
+    @abstractmethod
+    def warn(self, message: str, on: O):
+        pass
+    @abstractmethod
+    def error(self, message: str, on: O):
+        pass
+    @abstractmethod
+    def exceptionHook(self, exception: type[BaseException], value: BaseException, traceback):
+        pass
+
+
+class JsonLoggingStrategy(AbstractLoggingStrategy):
+    info_queue: list[M] = []
+    warning_queue: list[M] = []
+    error_queue: list[M] = []
+
+    @staticmethod
+    def _messageToJSON(m: M):
+        loglevel, message, obj = m
+        return {
+            "level": loglevel.name,
+            "message": message,
+            "object": (obj.toJSON() if isinstance(obj, AbstractNetboxType) else {
+                "type": "cosmo_string", "value": str(obj)
+            }),
+        }
+
+    def info(self, message: str, on: O):
+        self.info_queue.append((InfoLogLevel(), message, on))
+
+    def warn(self, message: str, on: O):
+        self.warning_queue.append((WarningLogLevel(), message, on))
+
+    def error(self, message: str, on: O):
+        self.error_queue.append((ErrorLogLevel(), message, on))
+
+    def flush(self):
+        # JSON-RPC like
+        res = {}
+        if len(self.warning_queue) + len(self.error_queue) == 0:
+            res = {
+                "result": list(map(self._messageToJSON, self.info_queue)),
+            }
+        else:
+            res = {
+                "error": list(map(self._messageToJSON, self.error_queue)),
+                "warning": list(map(self._messageToJSON, self.warning_queue)),
+            }
+        print(json.dumps(res))
+
+    def exceptionHook(self, exception: type[BaseException], value: BaseException, traceback):
+        if isinstance(exception, AbstractRecoverableError):
+            self.warn(str(value), None)
+        else:
+            self.error(str(value), None)
+
+
+class HumanReadableLoggingStrategy(AbstractLoggingStrategy):
+    def __init__(self, *args, netbox_instance_url: str, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.nb_instance_url = netbox_instance_url
+
+    def formatObject(self, obj: O):
```

Comment:
I am still not sold on the logging format.
Currently, it is a crime to punctuation.

I would go to a two-line-format to give us a bit more room:
```
[WARNING] [DSCL1234] [ifp-0/1/2] This is an invalid MTU, must not be 9191.
https://demo.netbox.dev/dcim/interface/12356

[LEVEL] ...
```

If we only have a device and not an object, just skip the third pair of braces.

I am open for better ideas, but I would like to get the links out of the text the user has to read..